client/state/posts/selectors.js

```diff
@@ -149,6 +151,38 @@ export function getSitePostsForQueryIgnoringPage( state, siteId, query ) {
 }
 
 /**
+ * Returns an array of posts for the posts query, including all known queried
+ * pages, preserving hierarchy. Returns null if no posts have been received.
+ * Hierarchy is represented by `parent` and `items` properties on each post.
+ *
+ * @param  {Object} state  Global state tree
+ * @param  {Number} siteId Site ID
+ * @param  {Object} query  Post query object
+ * @return {?Array}        Hierarchical posts for the post query
+ */
+export const getSitePostsHierarchyForQueryIgnoringPage = createSelector(
+	( state, siteId, query ) => {
+		let sitePosts = getSitePostsForQueryIgnoringPage( state, siteId, query );
+		if ( ! sitePosts ) {
+			return sitePosts;
+		}
+
+		// For each site post object, add `parent` and `order` properties.
+		// These are used by the TreeConvert library to build the hierarchy.
+		sitePosts = sitePosts.map( ( post, i ) => {
```

Comment:
Minor, but let's be a bit defensive here and use a different variable name. In longer functions we can mistake what sitePosts contains.
```
const sitePosts = getSitePostsForQueryIgnoringPage( state, siteId, query );
//...
const flatPosts = sitePosts.map( ( post, i ) => { //...
```